### DISCUSSION TITLE
WARNING: Lerna packages, TypeScript exporting, and Prisma clients are very tough to get working together

### DISCUSSION THREAD

**Discussion by HyperCrowd** [prisma/prisma#6981] [Q&A]
For those who are trying to use Lerna and have packages that have their own Prisma schemas and suffering from Google results being pure trash, read this!

1. Write a Lerna package (A) that exports classes and types
2. Write another Lerna package (B) that import's A's classes and types
3. Because Prisma strives for a "single instance", package A and package B have their own `schema.prisma` that connects to their own databases.
4. Each package has to generate it's own client.
5. To prevent multiple clients from existing in the same Lerna project needlessly, write a static singleton instance manager (map) for Prisma clients where the key is a hash of the connection string and the value is the Prisma client instance
6. Generate your Prisma clients for each package.
7. You suddenly discover each package's PrismaClient instance has its own unique type definitions and that there is no common PrismaClient type or interface:

```ts
// The default prisma client location has a specific type
import { PrismaClient } from '.prisma/client'

// Meanwhile, a custom prisma client  for a package with it's own schema.prisma has a type that is totally different than another package's prisma client.

import { PrismaClient } from '../prisma' 
```
8. You quickly realize your static singleton manager has to pass up the Prisma client type as a generic.
9. You discover your own interface for the PrismaClient proves to be a nightmare.
10. You quickly realize TypeScript doesn't play nice with class generics and statics, so you have to do some wild hacks to make it work.

So, for those who find themselves in this place: The best PrismaClient type for class generics and singleton instance management is `any` because you will experience a perfect combination of TypeScript shortcomings, Prisma design planning, and Lerna recommendations that will bring you endless misery.

Good luck!

**Comment by janpio** [comment]
Hmm, did you just want to (rightfully maybe!) rant about this or do you have a good suggestion? Do you maybe want to show us part of your code so we can see if we can do things better, or possibly change how we do things over at Prisma to make this less bad?

**Reply by HyperCrowd** [reply]
Howdy.

The SEO for this problem is pretty terrible, so I figured adding this warning in the issues would bring this warning to the attention of people searching for the dreaded combo of Lerna + TypeScript + Prisma.

The problems , I believe, are unfixable:

- Lerna fragments a project into packages, which creates the possibility of multiple Prisma clients of existing in the same project.
- Prisma is baking in the table schemas as types, so yes, the Prisma client types HAVE to be different.  This should probably be spelled somewhere in documentation.
- The need for only one Prisma client instance makes sense, as is covered in detail in the Prisma documentation.
- This creates a need for singleton instance management, but TypeScript cannot support class generics and static classes because you can't instantiate static classes.  This is logical, but alternatives to single instance managers/enforcers are not promoted outside of "_just export an instance instead of a class_" approach as many in the TypeScript camp believe singletons are evil and should be purged.

Each one of these problems touches a core design preference for each tool, and when put together, there isn't an intuitive way to get around each tools preference.

_Potential solution_: Prisma **could** assert an opinion on client instance management (complete with type exports!), but I haven't gotten to the point where multiple Prisma clients are live in production, many of which may connect to the same database, and what kind of traffic horror that might reveal.  Which is to say, relying on singletons to make the linter happy about passing a `PrismaClient` instance around, an instance whose types are autogenerated and trusted by the very reason that we are using Prisma at all, doesn't seem to be a valid reason to go about such a major course correction.

In fact, the more I use TypeScript, the more I realize how needlessly paranoid all of it's `strict` mode capabilities are.  The argument is always "_Types saves time later!_" which I agree with, but what I want to know is how much time each of these imposed rules from `strict` mode precisely saves.  How much time does `@typescript-eslint/prefer-function-type` really save in a refactor?  I refuse to believe all rules save equal amounts of time as each other rule.

I've learned to lie to my linter A LOT in TypeScript when dealing with Node modules anyways. :D  I think many TypeScript people have learned the fine art of fibbing to the linter.

Anyways, getting the typing of multiple `PrismaClient` instances to be compliant with a linter for what is essentially going to be a one-time singleton instance management use case ANYWAYS doesn't strike me as a priority for a major course correction, so in the meantime, people will stumble across these things and I hope this ticket pops up in their SEO journeys. :D

**Reply by janpio** [reply]
We solved more complicated problems in the past with getting Prisma to work with _any_ project setup, so I would be pretty happy if you could create a minimal reproduction that shows what you mean and just show it to us. We can at least try then.

**Reply by HyperCrowd** [reply]
I'll need some time to recreate a purely isolated and reproducible scenario which shows the problems I am detailing.

I appreciate your vigor and confidence, btw.  Hard to find when projects are overwhelmed by issue count. <3

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
